Linking Error when enable AddressSanitizer

**Describe the bug**
When enable AddressSanitizer, the linking will report an error.
```
$ python3 `which scons` build/RISCV/gem5.opt -j`nproc` --with-asan --without-tcmalloc```
```

```
scons: Reading SConscript files ...
Mkdir("/home/wangkaixuan/RantedWorkspace/pe-gem5/build/RISCV/gem5.build")
Checking for linker -Wl,--as-needed support... (cached) yes

Warning: To suppress false positive leaks, set the LSAN_OPTIONS environment variable to "suppressions=/home/wangkaixuan/RantedWorkspace/pe-gem5/util/lsan-suppressions:print_suppressions=0" when running gem5
Warning: LSAN_OPTIONS=suppressions=/home/wangkaixuan/RantedWorkspace/pe-gem5/util/lsan-suppressions:print_suppressions=0

Checking for compiler -gz support... (cached) yes
Checking for linker -gz support... (cached) yes
Info: Using Python config: python3-config
Checking for C header file Python.h... (cached) yes
Checking Python version... (cached) 3.10.12
Checking for accept(0,0,0) in C++ library None... (cached) yes
Checking for zlibVersion() in C++ library z... (cached) yes
Building in /home/wangkaixuan/RantedWorkspace/pe-gem5/build/RISCV
"build_tools/kconfig_base.py" "/home/wangkaixuan/RantedWorkspace/pe-gem5/build/RISCV/gem5.build/Kconfig" "/home/wangkaixuan/RantedWorkspace/pe-gem5/src/Kconfig"
Checking size of struct kvm_xsave ... (cached) yes
Checking for C header file fenv.h... (cached) yes
Checking for C header file png.h... (cached) yes
Checking for clock_nanosleep(0,0,NULL,NULL) in C library None... (cached) yes
Checking for C header file valgrind/valgrind.h... (cached) no
Checking for pkg-config package hdf5-serial... (cached) yes
Checking for H5Fcreate("", 0, 0, 0) in C library hdf5... (cached) yes
Checking for H5::H5File("", 0) in C++ library hdf5_cpp... (cached) yes
Checking for C header file capstone/capstone.h... (cached) no
Warning: Header file <capstone/capstone.h> not found.
         This host has no capstone library installed.
Checking for C header file linux/kvm.h... (cached) yes
Checking for timer_create(CLOCK_MONOTONIC, NULL, NULL) in C library None... (cached) yes
Checking for member exclude_host in struct perf_event_attr...(cached) yes
Checking for C header file linux/if_tun.h... (cached) yes
Checking for shm_open("/test", 0, 0) in C library None... (cached) yes
Checking for pkg-config package protobuf... (cached) yes
Checking for backtrace_symbols_fd((void *)1, 0, 0) in C library None... (cached) yes
Checking whether __i386__ is declared... (cached) no
Checking whether __x86_64__ is declared... (cached) yes
Checking for compiler -Wno-self-assign-overloaded support... (cached) yes
Checking for linker -Wno-free-nonheap-object support... (cached) yes
scons: done reading SConscript files.
scons: Building targets ...
 [VER TAGS]  -> RISCV/sim/tags.cc
 [    LINK]  -> RISCV/gem5.opt
`_ZN4gem58RiscvISAL14vecRegClassOpsE' referenced in section `.ldata.rel.local' of build/RISCV/arch/riscv/generated/inst-constrs.o: defined in discarded section `.data.rel.local._ZN4gem58RiscvISAL14vecRegClassOpsE[_ZN4gem58RiscvISA11vecRegClassE]' of build/RISCV/arch/riscv/generated/inst-constrs.o
`_ZN4gem58RiscvISAL14vecRegClassOpsE' referenced in section `.data.rel.local' of build/RISCV/arch/riscv/generated/generic_cpu_exec.o: defined in discarded section `.data.rel.local._ZN4gem58RiscvISAL14vecRegClassOpsE[_ZN4gem58RiscvISA11vecRegClassE]' of build/RISCV/arch/riscv/generated/generic_cpu_exec.o
`_ZN4gem58RiscvISAL14vecRegClassOpsE' referenced in section `.data.rel.local' of build/RISCV/arch/riscv/insts/vector.o: defined in discarded section `.data.rel.local._ZN4gem58RiscvISAL14vecRegClassOpsE[_ZN4gem58RiscvISA11vecRegClassE]' of build/RISCV/arch/riscv/insts/vector.o
collect2: error: ld returned 1 exit status
scons: *** [build/RISCV/gem5.opt] Error 1
scons: building terminated because of errors.
*** Summary of Warnings ***
Warning: To suppress false positive leaks, set the LSAN_OPTIONS environment variable to "suppressions=/home/wangkaixuan/RantedWorkspace/pe-gem5/util/lsan-suppressions:print_suppressions=0" when running gem5
Warning: LSAN_OPTIONS=suppressions=/home/wangkaixuan/RantedWorkspace/pe-gem5/util/lsan-suppressions:print_suppressions=0
Warning: Header file <capstone/capstone.h> not found.
         This host has no capstone library installed.
```

**Affects version**
V23-1

**gem5 Modifications**
No Modification

I have ran various tests and found that the incompatibility is with GCC and using the AddressSanitizer. Clang works just fine.

Though this is somewhat dismissive, I believe this bug is symptomatic of some complicated code we have in our codebase. For example:

```
`_ZN4gem58RiscvISAL14vecRegClassOpsE' referenced in section `.data.rel.local' of build/RISCV/arch/riscv/insts/vector.o: defined in discarded section `.data.rel.local._ZN4gem58RiscvISAL14vecRegClassOpsE[_ZN4gem58RiscvISA11vecRegClassE]' of build/RISCV/arch/riscv/insts/vector.o
```

appears (to me at least) to be getting confused by this: https://github.com/gem5/gem5/blob/e8bc4fc137a5a7e82b601432271a027b652ae69b/src/arch/riscv/regs/vector.hh#L69 

I don't even know what the `static inline` does here exactly...


I have created #951 to disable the Address Sanitizer for GCC. This doesn't fix the bug per-se but will stop users accidentally running into it.

**Note**: I also found the Undefined Behavior Sanitizer was failing in GCC with the following error (when compiling using `scons build/ALL/gem5.opt -j`nproc` --with-ubsan`):

```sh
[     CXX] src/base/loader/image_file_data.cc -> ALL/base/loader/image_file_data.o
In file included from /usr/include/string.h:535,
                 from /usr/include/c++/11/cstring:42,
                 from src/base/cprintf_formats.hh:33,
                 from src/base/cprintf.hh:38,
                 from src/base/logging.hh:49,
                 from src/base/loader/image_file_data.cc:40:
In function ‘char* strcpy(char*, const char*)’,
    inlined from ‘int gem5::loader::doGzipLoad(int)’ at src/base/loader/image_file_data.cc:70:11,
    inlined from ‘gem5::loader::ImageFileData::ImageFileData(const string&)’ atsrc/base/loader/image_file_data.cc:116:24:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:79:33: error: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ offset [0, 19] is out of the bounds [0, 0] [-Werror=array-bounds]
   79 |   return __builtin___strcpy_chk (__dest, __src, __glibc_objsize (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
scons: *** [build/ALL/base/loader/image_file_data.o] Error 1
scons: building terminated because of errors.
```

In similar vein, I used PR #951 to remove this for GCC. I don't know exactly what is causing the problem.

--- a/SConstruct
+++ b/SConstruct
@@ -707,20 +707,26 @@ for variant_path in variant_paths:
     if GetOption('with_ubsan'):
         sanitizers.append('undefined')
     if GetOption('with_asan'):
-        # Available for gcc >= 5 or llvm >= 3.1 both a requirement
-        # by the build system
-        sanitizers.append('address')
-        suppressions_file = Dir('util').File('lsan-suppressions').get_abspath()
-        suppressions_opt = 'suppressions=%s' % suppressions_file
-        suppressions_opts = ':'.join([suppressions_opt,
-                                      'print_suppressions=0'])
-        env['ENV']['LSAN_OPTIONS'] = suppressions_opts
-        print()
-        warning('To suppress false positive leaks, set the LSAN_OPTIONS '
-                'environment variable to "%s" when running gem5' %
-                suppressions_opts)
-        warning('LSAN_OPTIONS=%s' % suppressions_opts)
-        print()
+        if env['GCC']:
+            # Address sanitizer is not supported with GCC. Please see Github
+            # Issue https://github.com/gem5/gem5/issues/916 for more details.
+            warning("Address Sanitizer is not supported with GCC. "
+                    "This option will be ignored.")
+        else:
+            # Available for llvm >= 3.1. A requirement by the build system.
+            sanitizers.append('address')
+            suppressions_file = Dir('util').File('lsan-suppressions')\
+                                .get_abspath()
+            suppressions_opt = 'suppressions=%s' % suppressions_file
+            suppressions_opts = ':'.join([suppressions_opt,
+                                        'print_suppressions=0'])
+            env['ENV']['LSAN_OPTIONS'] = suppressions_opts
+            print()
+            warning('To suppress false positive leaks, set the LSAN_OPTIONS '
+                    'environment variable to "%s" when running gem5' %
+                    suppressions_opts)
+            warning('LSAN_OPTIONS=%s' % suppressions_opts)
+            print()
     if sanitizers:
         sanitizers = ','.join(sanitizers)
         if env['GCC'] or env['CLANG']:

